### PR TITLE
12-add-listing-package

### LINKED ISSUE #12: Add listing class
Add `Listing` class to represent job listing

### PR DESCRIPTION
**Pull request by clevon-w**
## Which ticket does this address

- closes #12

## What problems does this solve

- Add Listing class

## Solution/Changes

- Added listing package which contains Listing, Title, Description classes

## Checklist

- [x] All tests pass locally and on CI

### CODE REVIEW

File: src/main/java/seedu/address/model/listing/Description.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
+package seedu.address.model.listing;
+
+import static java.util.Objects.requireNonNull;
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+/**
+ * Represents a Listing's description in GoodMatch.
+ * Guarantees: immutable; is valid as declared in {@link #isValidDescription(String)}
+ */
+public class Description {
```
Comment (yyj-02) [inline src/main/java/seedu/address/model/listing/Description.java:10]: Can chang to JobDescription

File: src/test/java/seedu/address/testutil/ListingBuilder.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
+package seedu.address.testutil;
+
+import java.util.ArrayList;
+
+import seedu.address.model.listing.Description;
+import seedu.address.model.listing.Listing;
+import seedu.address.model.listing.Title;
+
+/**
+ * A utility class to help with building Person objects.
```
Comment (yyj-02) [inline src/test/java/seedu/address/testutil/ListingBuilder.java:10]: *Listingobjects

File: src/test/java/seedu/address/testutil/TypicalListings.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+import seedu.address.model.listing.Listing;
+
+
+/**
+ * A utility class containing a list of {@code Person} objects to be used in tests.
+ */
+public class TypicalListings {
+
+    public static final Listing CHICKENRICEUNCLE = new ListingBuilder()
```
Comment (yyj-02) [inline src/test/java/seedu/address/testutil/TypicalListings.java:19]: Change to UPPER_SNAKE_CASE

**Review by yyj-02 — CHANGES_REQUESTED** [review]
Just need some minor refactor/edits, looks good overall

File: src/test/java/seedu/address/testutil/TypicalListings.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import seedu.address.model.listing.Listing;
+
+
+/**
+ * A utility class containing a list of {@code Person} objects to be used in tests.
```
Comment (homuntan02) [inline src/test/java/seedu/address/testutil/TypicalListings.java:15]: Change comment to {@code Listing}

File: src/main/java/seedu/address/model/listing/Listing.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            return true;
+        }
+
+        return otherListing != null
+                && otherListing.getTitle().equals(getTitle());
+    }
+
+    /**
+     * Returns true if both persons have the same identity and data fields.
+     * This defines a stronger notion of equality between two persons.
```
Comment (homuntan02) [inline src/main/java/seedu/address/model/listing/Listing.java:63]: Change comment to listings instead of persons

File: src/test/java/seedu/address/model/listing/ListingTest.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        // same object -> returns true
+        assertTrue(CHICKENRICEUNCLE.equals(CHICKENRICEUNCLE));
+
+        // null -> returns false
+        assertFalse(CHICKENRICEUNCLE.equals(null));
+
+        // different type -> returns false
+        assertFalse(CHICKENRICEUNCLE.equals(5));
+
+        // different person -> returns false
```
Comment (homuntan02) [inline src/test/java/seedu/address/model/listing/ListingTest.java:58]: change to different listing

File: src/test/java/seedu/address/model/listing/ListingTest.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+import static seedu.address.testutil.TypicalListings.CHICKENRICEUNCLE;
+import static seedu.address.testutil.TypicalListings.TOILETCLEANER;
+
+import org.junit.jupiter.api.Test;
+
+import seedu.address.testutil.ListingBuilder;
+
+public class ListingTest {
+    @Test
+    public void isSamePerson() {
```
Comment (homuntan02) [inline src/test/java/seedu/address/model/listing/ListingTest.java:17]: Method name change to isSameListing

File: src/test/java/seedu/address/testutil/ListingBuilder.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            "This is a default job description! "
+            + "In this job you are expected to do these things: thing1, thing2 and thing3.";
+    public static final ArrayList<String> DEFAULT_APPLICANTS = new ArrayList<>();
+
+    private Title title;
+    private Description description;
+    private ArrayList<String> applicants;
+
+    /**
+     * Creates a {@code PersonBuilder} with the default details.
```
Comment (homuntan02) [inline src/test/java/seedu/address/testutil/ListingBuilder.java:25]: Change to {@code ListingBuilder}

**Review by homuntan02 — CHANGES_REQUESTED** [review]

File: src/main/java/seedu/address/model/listing/Description.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+     */
+    public Description(String description) {
+        requireNonNull(description);
+        description = description.strip();
+        checkArgument(isValidDescription(description), MESSAGE_CONSTRAINTS);
+        fullDescription = description;
+    }
+
+    /**
+     * Returns true if a given string is a valid description.
```
[the remark this answers is not in the document]
Comment (clevon-w) [reply src/main/java/seedu/address/model/listing/Description.java:37]: Added.

File: src/main/java/seedu/address/model/listing/Title.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
+package seedu.address.model.listing;
+
+import static java.util.Objects.requireNonNull;
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+/**
+ * Represents a Listing's title in GoodMatch.
+ * Guarantees: immutable; is valid as declared in {@link #isValidTitle(String)}
+ */
+public class Title {
```
[the remark this answers is not in the document]
Comment (clevon-w) [reply src/main/java/seedu/address/model/listing/Title.java:10]: changed to JobTitle

File: src/main/java/seedu/address/model/listing/Description.java (hunk 1 above)
Comment (clevon-w) [reply src/main/java/seedu/address/model/listing/Description.java:10]: renamed

File: src/test/java/seedu/address/testutil/TypicalListings.java (hunk 3 above)
Comment (clevon-w) [reply src/test/java/seedu/address/testutil/TypicalListings.java:19]: Updated this and also for toilet cleaner and software developer

File: src/test/java/seedu/address/testutil/TypicalListings.java (hunk 4 above)
Comment (clevon-w) [reply src/test/java/seedu/address/testutil/TypicalListings.java:15]: updated

File: src/test/java/seedu/address/model/listing/ListingTest.java (hunk 7 above)
Comment (clevon-w) [reply src/test/java/seedu/address/model/listing/ListingTest.java:17]: changed

File: src/test/java/seedu/address/model/listing/ListingTest.java (hunk 6 above)
Comment (clevon-w) [reply src/test/java/seedu/address/model/listing/ListingTest.java:58]: updated this and the comments for the different title, description and applicants

File: src/main/java/seedu/address/model/listing/Listing.java (hunk 5 above)
Comment (clevon-w) [reply src/main/java/seedu/address/model/listing/Listing.java:63]: updated

File: src/test/java/seedu/address/testutil/ListingBuilder.java (hunk 8 above)
Comment (clevon-w) [reply src/test/java/seedu/address/testutil/ListingBuilder.java:25]: updated all PersonBuilder instances to ListingBuilder

File: src/test/java/seedu/address/testutil/ListingBuilder.java (hunk 2 above)
Comment (clevon-w) [reply src/test/java/seedu/address/testutil/ListingBuilder.java:10]: updated

File: docs/AboutUs.md [hunk 11]
```diff
@@ -1,6 +1,6 @@
 ---
 layout: page
-title: About Us
```
Comment (yyj-02) [inline docs/AboutUs.md:3]: I don't think the change in these markdown files is intended right?

File: docs/_includes/header.html [hunk 12]
```diff
@@ -3,8 +3,8 @@
   <div class="wrapper">
     {%- assign default_paths = site.pages | map: "path" -%}
     {%- assign page_paths = site.header_pages | default: default_paths -%}
-    {%- assign titles_size = site.pages | map: 'title' | join: '' | size -%}
```
Comment (yyj-02) [inline docs/_includes/header.html:6]: Are these changes intended?

File: docs/_layouts/page.html [hunk 13]
```diff
@@ -4,7 +4,7 @@
 <article class="post">
 
   <header class="post-header">
-    <h1 class="post-title">{{ page.title | escape }}</h1>
```
Comment (yyj-02) [inline docs/_layouts/page.html:7]: The html pages here too, might be unintended

File: src/main/resources/view/HelpWindow.fxml [hunk 14]
```diff
@@ -9,7 +9,7 @@
 <?import javafx.scene.layout.HBox?>
 <?import javafx.stage.Stage?>
 
-<fx:root resizable="false" title="Help" type="javafx.stage.Stage" xmlns="http://javafx.com/javafx/11" xmlns:fx="http://javafx.com/fxml/1">
```
Comment (yyj-02) [inline src/main/resources/view/HelpWindow.fxml:12]: ^

File: src/main/resources/view/MainWindow.fxml [hunk 15]
```diff
@@ -12,7 +12,7 @@
 <?import javafx.scene.layout.VBox?>
 
 <fx:root type="javafx.stage.Stage" xmlns="http://javafx.com/javafx/8" xmlns:fx="http://javafx.com/fxml/1"
-         title="Address App" minWidth="450" minHeight="600" onCloseRequest="#handleExit">
```
Comment (yyj-02) [inline src/main/resources/view/MainWindow.fxml:15]: ^

**Review by yyj-02 — CHANGES_REQUESTED** [review]
I think you have committed quite a few unintended changes when renaming `title` to `jobTitle`

**Review by yyj-02 — APPROVED** [review]
LGTM

**Review by homuntan02 — APPROVED** [review]
Looking good
